packages/access-api/src/models/delegations.js

```diff
@@ -0,0 +1,137 @@
+import * as Ucanto from '@ucanto/interface'
+import {
+  delegationsToBytes,
+  bytesToDelegations,
+} from '@web3-storage/access/encoding'
+
+/**
+ * @typedef {import('@web3-storage/access/src/types').DelegationTable} DelegationRow
+ * @typedef {Omit<DelegationRow, 'inserted_at'|'updated_at'|'expires_at'>} DelegationRowUpdate
+ */
+
+/**
+ * @typedef Tables
+ * @property {DelegationRow} delegations
+ */
+
+/**
+ * @typedef {import("../types/database").Database<Tables>} DelegationsDatabase
+ */
+
+/**
+ * DelegationsStorage that persists using SQL.
+ * * should work with cloudflare D1
+ */
+export class DbDelegationsStorage {
+  /** @type {DelegationsDatabase} */
+  #db
+
+  /**
+   * @param {DelegationsDatabase} db
+   */
+  constructor(db) {
+    this.#db = db
+    // eslint-disable-next-line no-void
+    void (
+      /** @type {import('../types/delegations').DelegationsStorage} */ (this)
+    )
+  }
+
+  get length() {
+    return new Promise((resolve, reject) => {
+      this.#db
+        .selectFrom('delegations')
+        .select((e) => e.fn.count('cid').as('size'))
+        .executeTakeFirstOrThrow()
+        .then(({ size }) => {
+          if (typeof size === 'string') {
+            const sizeNumber = parseInt(size, 10)
+            if (isNaN(sizeNumber)) {
+              throw new TypeError(
+                `unable to determine size number of delegations table`
+              )
+            }
+            return sizeNumber
+          }
+          if (typeof size === 'bigint') {
+            if (size > Number.MAX_SAFE_INTEGER) {
+              throw new TypeError(`table size too big for js Number`)
+            }
+            return Number(size)
+          }
+          return size
+        })
+        .then(resolve)
+        .catch(reject)
+    })
+  }
+
+  /**
+   * store items
+   *
+   * @param  {Array<Ucanto.Delegation>} delegations
+   * @returns {Promise<void>}
+   */
+  push = async (...delegations) => {
+    if (delegations.length === 0) {
+      return
+    }
+    const values = delegations.map((d) => createDelegationRowUpdate(d))
+    await this.#db
+      .insertInto('delegations')
+      .values(values)
+      .onConflict((oc) => oc.column('cid').doNothing())
+      .executeTakeFirst()
+  }
+
+  /**
+   * iterate through all stored items
+   *
+   * @returns {AsyncIterableIterator<Ucanto.Delegation>}
+   */
+  async *[Symbol.asyncIterator]() {
+    if (this.#db.canStream) {
```

Comment:
We should not allow this if the database cannot stream.

Reply:
Why not? Would it be sufficient to e.g. only refuse to do that in production or if the overall db size is very large.

I removed the non-streaming case for now, but it was needed for a WIP version of the `testDelegateThenClaim` test. https://github.com/web3-storage/w3protocol/actions/runs/4256209423/jobs/7404788157#step:7:48
But for now I've removed it and will just disable tests that relied on it (since further work #394 will lead to a better impl that doesn't need asyncIteration).

Reply:
Hmm so at the time I didn't know that this was going to become a filtered list - in the case of all data then there's no point in allowing buffered response as it'll immediately become unusable.

For a specific DID's delegations then sure buffered response is fine.